oasis-engine/engine
# Test: Use Floss and Chai to Replace Jest

## Advantages

- Much more easier to write your unit test
- Much more easier to debug your unit test



## Packages Added

- [Floss](https://github.com/pixijs/floss) is a unit-testing tool made by pixi team which help us run our test in electron.

- [Chai](https://www.chaijs.com/api/bdd/) is an assertion library for [[node](http://nodejs.org/)](http://nodejs.org/) and the browser.

- Use [nyc](https://github.com/istanbuljs/nyc) to collect coverage.



## How to Write Unit Test

1. Add unit test in tests packages
2. Add unit test dependencies in `package.json` in tests
3. Use `npx lerna boostrap` update dependencies.



A Demo for unit test: 

```typescript
import { expect } from "chai";
import { WebGLEngine, WebCanvas } from "@oasis-engine/rhi-webgl";

describe("webgl engine test", () => {
  it("create a webgl engine", () => {
    const canvas = document.createElement("canvas");
    const engine = new WebGLEngine(canvas);
    expect(engine).not.be.null;
  });
});
```

Chai is quite different with Jest. We need migrate it.



## How to Test

#### Test

```shell
npm run test
```



<img src="https://gw.alipayobjects.com/zos/OasisHub/edeb5b4a-d114-40a4-9a97-7b029a00adb9/image-20220430183134839.png" alt="image-20220430183134839" style="zoom:50%;" />



#### Test Coverage

```shell
npm run test-cov
```

And the coverage page is:

<img src="https://gw.alipayobjects.com/zos/OasisHub/6bde1760-52c2-494a-9d4a-a15b5c2e66c7/image-20220430183443304.png" alt="image-20220430183443304" style="zoom:50%;" />



### Test Debug

If you want to debug your coverage, you should run:

```shell
npm run test-debug
```

Test tool will open electron app and debug tool:

<img src="https://gw.alipayobjects.com/zos/OasisHub/2a0c3559-f8b4-4286-9ace-35c250463888/image-20220430183706093.png" alt="image-20220430183706093" style="zoom:50%;" />



### Error Handling

Run test

<img src="https://gw.alipayobjects.com/zos/OasisHub/0c2644bd-ffb8-45b9-a622-29aa1cc28a51/image-20220506002944946.png" alt="image-20220506002944946" style="zoom:50%;" /> 

You can use `debugger` in `npm run test-debug`



<img src="https://gw.alipayobjects.com/zos/OasisHub/04b5c9e4-ba65-42f8-a785-3a3a00009962/image-20220506003144352.png" alt="image-20220506003144352" style="zoom:50%;" />
